[view-transitions] Register user agent `@keyframes` rules

nt1m opened the pull request:
#### 305495110963caa7a6ea92bf96631017d9bae4e7
<pre>
[view-transitions] Register user agent `@keyframes` rules
<a href="https://bugs.webkit.org/show_bug.cgi?id=269504">https://bugs.webkit.org/show_bug.cgi?id=269504</a>
<a href="https://rdar.apple.com/123033600">rdar://123033600</a>

Reviewed by Matthieu Dubet.

Add static storage for user agent @keyframes, and use it for the viewTransitions.css stylesheet.

This is the same pattern used as `CounterStyleRegistry::userAgentCounterStyles`.

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::userAgentKeyframes):
(WebCore::Style::Resolver::addUserAgentKeyframeStyle):
(WebCore::Style::Resolver::isAnimationNameValid):
(WebCore::Style::Resolver::keyframeRulesForName const):
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::addUserAgentKeyframes):
(WebCore::Style::UserAgentStyle::ensureDefaultStyleSheetsForElement):

Canonical link: <a href="https://commits.webkit.org/274789@main">https://commits.webkit.org/274789@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c4c2badea3260d065239f732c3867faf8c1591a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40030 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42406 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42575 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42337 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21951 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16371 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40604 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16031 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34584 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43853 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36319 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12189 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/16464 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/8987 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16108 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

mdubet commented on Source/WebCore/style/StyleResolver.cpp:
```diff
@@ -216,6 +216,18 @@ void Resolver::appendAuthorStyleSheets(const Vector<RefPtr<CSSStyleSheet>>& styl
+
+void Resolver::addUserAgentKeyframeStyle(Ref<StyleRuleKeyframes>&& rule)
+{
+    auto& animationName = rule->name();
```
`const auto&`?

mdubet replied:
Or just get rid of the local variable completely

mdubet commented on Source/WebCore/style/UserAgentStyle.cpp:
```diff
@@ -130,6 +132,14 @@ void static addToCounterStyleRegistry(StyleSheetContents& sheet)
 
+void static addUserAgentKeyframes(StyleSheetContents& sheet)
+{
+    for (auto& rule : sheet.childRules()) {
```
What about nested keyframes rule? It seems this will only get top level ones ?

mdubet reviewed (approved):
The `static` makes me a little worried 😅

nt1m committed:
[view-transitions] Register user agent `@keyframes` rules
https://bugs.webkit.org/show_bug.cgi?id=269504
rdar://123033600

Reviewed by Matthieu Dubet.

Add static storage for user agent @keyframes, and use it for the viewTransitions.css stylesheet.

This is the same pattern used as `CounterStyleRegistry::userAgentCounterStyles`.

* Source/WebCore/style/StyleResolver.cpp:
(WebCore::Style::Resolver::userAgentKeyframes):
(WebCore::Style::Resolver::addUserAgentKeyframeStyle):
(WebCore::Style::Resolver::isAnimationNameValid):
(WebCore::Style::Resolver::keyframeRulesForName const):
* Source/WebCore/style/StyleResolver.h:
* Source/WebCore/style/UserAgentStyle.cpp:
(WebCore::Style::addUserAgentKeyframes):
(WebCore::Style::UserAgentStyle::ensureDefaultStyleSheetsForElement):

Canonical link: https://commits.webkit.org/274789@main
